### PR TITLE
[Fix]Guest Authentication Flow

### PR DESCRIPTION
**Pull request by ipavlidakis**
### 🔗 Issue Links

Resolves https://github.com/GetStream/ios-issues-tracking/issues/389

### 🎯 Goal

Fix the Guest authentication

### 📝 Summary

Since the SDK version 4.25.0 the Guest Authentication was failing. This PR resolves the underlying issue and ensures Guest Authentication works again as expected.

### 🛠 Implementation

During Guest Authentication, the APIClient's operationQueue is getting suspended as it enters tokenFetching mode. Then we using APIClient to fetch the Guest Authentication Token. The request never gets executed due to the suspended operationQueue.

To overcome the issue but also provide a mechanism for similar use-cases, we are introducing a new request/operation combination on APIClient. The new request will be translated to an AsyncOperation which will be added for execution to user-provided operationQueue. The APIClient will not be responsible for the execution of the new UnmanagedOperation.

### 🧪 Manual Testing Notes

From the DemoApp, try to connect with a Guest User

### ☑️ Contributor Checklist

- [X] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [X] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [X] Changelog is updated with client-facing changes
- [X] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

_Provide a funny gif or image that relates to your work on this pull request. (Optional)_

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Fix an issues which was causing the app to terminate when using a filter with the `in` operator and `cid` values [#2561](https://github.com/GetStream/stream-chat-swift/pull/2561)
 - Fix unexpected 401s produced at launch while the chat is not yet fully connected [#2559](https://github.com/GetStream/stream-chat-swift/pull/2559)
 - Fix crash when getting unread count in an invalid state [#2570](https://github.com/GetStream/stream-chat-swift/pull/2570)
+- Fix an issue which was blocking a Guest Authentication operation to retrieve a connection token [#2574](https://github.com/GetStream/stream-chat-swift/pull/2574)
```
Comment (nuno-vieira) [inline CHANGELOG.md:14]: In the changelog, we should focus on what is the impact of the customer, and not in the underlying implementation 👍 Here is a suggestion to improve it:
```suggestion
- Fix an issue where connecting a guest user would never call the completion block [#2574](https://github.com/GetStream/stream-chat-swift/pull/2574)
```

**Review by nuno-vieira — COMMENTED** [review]
The fix does not seem to work everytime 🤔 Especially when I open the app for the first time, the request to the channel list seems to fail:

image

Log: 

> 🚨 2023-04-18 14:33:21.554 [ERROR] [com.apple.NSURLSession-delegate] [APIClient.swift:223] [executeRequest(endpoint:completion:)] > Error MissingToken in /Users/nunovieira/Documents/Work/StreamChat/stream-chat-swift/Sources/StreamChat/APIClient/RequestEncoder.swift:142 -> Failed to get `token`, request can't be created.
⚠️ 2023-04-18 14:33:21.560 [WARNING] [com.apple.NSURLSession-delegate] [ConnectionRepository.swift:109] [disconnect(source:completion:)] > The client is already disconnected. Skipping the `disconnect` call.

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 2]
```diff
@@ -101,12 +101,34 @@ class APIClient {
...
     private func operation<Response: Decodable>(
         endpoint: Endpoint<Response>,
         isRecoveryOperation: Bool,
         completion: @escaping (Result<Response, Error>) -> Void
     ) -> AsyncOperation {
         AsyncOperation(maxRetries: maximumRequestRetries) { [weak self] operation, done in
+            guard self?.isRefreshingToken == false else {
+                completion(.failure(ClientError.RefreshingToken()))
+                return
+            }
```
Comment (polqf) [inline Sources/StreamChat/APIClient/APIClient.swift:130]: There are 2 problems here:
1. We are not completing the operation (you would need to call `done(.continue)`)
2. We should not be returning this error upwards, because otherwise the whole retry/queueing strategy is lost

Why did you move the logic from `executeRequest` to here?

File: Sources/StreamChat/Repositories/AuthenticationRepository.swift [hunk 3]
```diff
@@ -121,6 +121,16 @@ class AuthenticationRepository {
         self.currentUserId = currentUserId
     }
 
+    /// Fetches the user saved in the database, if exists
+    func deleteCurrentUser() {
+        if let currentUserId = currentUserId {
```
Comment (polqf) [inline Sources/StreamChat/Repositories/AuthenticationRepository.swift:126]: guard let?

File: StreamChatUITestsApp/ViewController.swift [hunk 4]
```diff
@@ -26,6 +26,7 @@ final class ViewController: UIViewController {
         stackView.distribution = .fillProportionally
         stackView.translatesAutoresizingMaskIntoConstraints = false
         stackView.addArrangedSubview(createStartButton())
+        stackView.addArrangedSubview(createConnectGuestButton())
```
Comment (polqf) [inline StreamChatUITestsApp/ViewController.swift:29]: 👏

File: Sources/StreamChat/Utils/Operations/AsyncOperation.swift [hunk 5]
```diff
@@ -56,6 +56,8 @@ class AsyncOperation: BaseOperation {
     }
 }
 
+class UnmanagedAsyncOperation: AsyncOperation {}
```
Comment (polqf) [inline Sources/StreamChat/Utils/Operations/AsyncOperation.swift:59]: Why is that needed?

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 6]
```diff
@@ -101,12 +101,34 @@ class APIClient {
...
+    /// Performs a network request and retries in case of network failures. The network operation
+    /// won't be managed by the APIClient instance. Instead it will be added on the provided operationQueue.
+    ///
+    /// - Parameters:
+    ///   - endpoint: The `Endpoint` used to create the network request.
+    ///   - operationQueue: The queue that will be responsible for executing the network operation
+    ///   - completion: Called when the networking request is finished.
+    func unmanagedRequest<Response: Decodable>(
+        endpoint: Endpoint<Response>,
+        operationQueue: OperationQueue,
```
Comment (polqf) [inline Sources/StreamChat/APIClient/APIClient.swift:113]: Given the usage, I think we can safely ignore this parameter and use .main

Comment (polqf) [reply Sources/StreamChat/APIClient/APIClient.swift:113]: I would also rename it to `priorityRequest`

File: Sources/StreamChat/APIClient/APIClient.swift (hunk 2 above)
Comment (ipavlidakis) [reply Sources/StreamChat/APIClient/APIClient.swift]: Great point! I updated it

File: Sources/StreamChat/APIClient/APIClient.swift (hunk 6 above)
Comment (ipavlidakis) [reply Sources/StreamChat/APIClient/APIClient.swift:113]: I'm a fan of `priorityRequest`. We aren't really applying any priority logic, we are just enqueuing in another queue. Unmanaged seems to be closer in accordance to what it does. Additionally it's also aligned with the Swift's Unmanaged API for manual (memory) management

**Review by polqf — APPROVED** [review]
I would only reword the Changelog to be more customer focused

File: StreamChatUITestsApp/ViewController.swift [hunk 7]
```diff
@@ -163,6 +171,15 @@ extension ViewController {
...
     }
 
+    func createConnectGuestButton() -> UIButton {
+        let startButton = UIButton(type: .system)
+        startButton.translatesAutoresizingMaskIntoConstraints = false
+        startButton.setTitle("Connect Guest", for: .normal)
+        startButton.accessibilityIdentifier = "TestApp.ConnectGuest"
+        startButton.addTarget(self, action: #selector(didTapConnectGuest), for: .touchUpInside)
+        return startButton
+    }
```
Comment (nuno-vieira) [inline StreamChatUITestsApp/ViewController.swift:181]: Why is this implemented? 🤔 I can't see any E2E Test which uses this?

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅ Just curious on why we implemented the Guest Button in the E2E App 🤔 It is to prepare the E2E Test?

**Comment by ipavlidakis** [comment]
@polqf Any suggestions on how you would write the Changelog?

**Comment by ipavlidakis** [comment]
> LGTM! ✅ Just curious on why we implemented the Guest Button in the E2E App 🤔 It is to prepare the E2E Test?

You are absolutely right. The changes are to prepare for E2E tests
